indico/migrations/versions/20231222_1408_b37cbc4bb129_link_events_to_reservation_occurrences.py

```diff
@@ -0,0 +1,121 @@
+        sa.PrimaryKeyConstraint('id'),
+        sa.ForeignKeyConstraint(['contribution_id'], ['events.contributions.id']),
+        sa.ForeignKeyConstraint(['event_id'], ['events.events.id']),
+        sa.ForeignKeyConstraint(['linked_event_id'], ['events.events.id']),
+        sa.ForeignKeyConstraint(['session_block_id'], ['events.session_blocks.id']),
+        sa.CheckConstraint('(event_id IS NULL) = (link_type = 1)', name='valid_event_id'),
+        sa.CheckConstraint('link_type != 2 '
+                           'OR (contribution_id IS NULL AND session_block_id IS NULL AND linked_event_id IS NOT NULL)',
+                           name='valid_event_link'),
+        sa.CheckConstraint('link_type != 3 '
+                           'OR (linked_event_id IS NULL AND session_block_id IS NULL AND contribution_id IS NOT NULL)',
+                           name='valid_contribution_link'),
+        sa.CheckConstraint('link_type != 6 '
+                           'OR (contribution_id IS NULL AND linked_event_id IS NULL AND session_block_id IS NOT NULL)',
+                           name='valid_session_block_link'),
+        schema='roombooking'
+    )
+    op.add_column('reservation_occurrences', sa.Column('link_id', sa.Integer(), nullable=True), schema='roombooking')
+    op.create_foreign_key(None, 'reservation_occurrences', 'reservation_occurrence_links', ['link_id'], ['id'],
+                          source_schema='roombooking', referent_schema='roombooking')
+    op.create_index(None, 'reservation_occurrence_links', ['contribution_id'], unique=False, schema='roombooking')
+    op.create_index(None, 'reservation_occurrence_links', ['event_id'], unique=False, schema='roombooking')
+    op.create_index(None, 'reservation_occurrence_links', ['linked_event_id'], unique=False, schema='roombooking')
+    op.create_index(None, 'reservation_occurrence_links', ['session_block_id'], unique=False, schema='roombooking')
+    op.create_index(None, 'reservation_occurrences', ['link_id'], unique=False, schema='roombooking')
+
+    # Migrate data
+    # TODO
+
+    # Drop reservation links
+    op.drop_index('ix_reservations_link_id', table_name='reservations', schema='roombooking')
+    op.drop_constraint('fk_reservations_link_id_reservation_links', 'reservations', schema='roombooking')
+    op.drop_table('reservation_links', schema='roombooking')
+    op.drop_column('reservations', 'link_id', schema='roombooking')
+
+
+def downgrade():
+    # Recreate reservation links
+    op.add_column('reservations', sa.Column('link_id', sa.INTEGER(), nullable=True),
+                  schema='roombooking')
+    op.create_foreign_key(None, 'reservations', 'reservation_links', ['link_id'], ['id'],
+                          source_schema='roombooking', referent_schema='roombooking')
+    op.create_table(
+        'reservation_links',
+        sa.Column('id', sa.INTEGER(), nullable=False),
+        sa.Column('link_type', sa.SMALLINT(), nullable=False),
```

Comment:
Can/should probably use the `PyIntEnum` type (downgrade is generated via reflection, so it uses native PG types instead of the usual SA types). Benefit: You can also remove lots of noise below (the enum-related constraints for example)

Reply:
Does that mean that I don't need to explicitly define the `valid_event_link`, `valid_contribution_link` and `valid_session_block_link` CHECK constraints when creating the `reservation_occurrence_links` table either?

Reply:
iirc those are coming from the link mixin so you'll still need them here.

or.... just rename the table and avoid all that trouble ;)

1. rename the table 
2. add the occurrence id column
3. migrate data
4. delete the reservation id column